src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Extension.cs

```diff
@@ -63,6 +64,321 @@ internal override string ExtensionName
+            {
+                // Note: using slash for nested type to match CIL: ECMA-335 I.10.7.2
+                if (namedType.ContainingType is { } containingType)
+                {
+                    appendContainingType(containingType, builder);
+                    builder.Append(containingType.Name);
+                    appendArity(containingType, builder);
+                    builder.Append('/');
+                }
+            }
+
+            static void appendArity(NamedTypeSymbol namedType, StringBuilder builder)
+            {
+                if (namedType.Arity > 0)
+                {
+                    builder.Append('`');
+                    builder.Append(StringExtensions.GetNumeral(namedType.Arity));
+                }
+            }
+
+            static void appendTypeArguments(NamedTypeSymbol namedType, StringBuilder builder)
+            {
+                var typeArguments = ArrayBuilder<TypeWithAnnotations>.GetInstance();
+                namedType.GetAllTypeArgumentsNoUseSiteDiagnostics(typeArguments);
+                if (typeArguments.Count > 0)
+                {
+                    builder.Append('<');
+                    for (int i = 0; i < typeArguments.Count; i++)
+                    {
+                        if (i > 0)
+                        {
+                            builder.Append(", ");
+                        }
+
+                        appendType(typeArguments[i].Type, builder);
+                    }
+
+                    builder.Append('>');
+                }
+
+                typeArguments.Free();
+            }
+
+            static void appendTypeParameter(TypeParameterSymbol typeParameter, StringBuilder builder, bool isDefinition)
+            {
+                if (isDefinition)
+                {
+                    var cciTypeParameter = (Cci.IGenericParameter)typeParameter.GetCciAdapter();
+                    Debug.Assert(cciTypeParameter.Variance == Cci.TypeParameterVariance.NonVariant);
+
+                    if (cciTypeParameter.MustBeReferenceType)
+                    {
+                        builder.Append("class ");
+                    }
+                    else if (cciTypeParameter.MustBeValueType)
+                    {
+                        builder.Append("valuetype ");
+                    }
+
+                    if (cciTypeParameter.AllowsRefLikeType)
+                    {
+                        builder.Append("byreflike ");
+                    }
+
+                    if (cciTypeParameter.MustHaveDefaultConstructor)
+                    {
+                        builder.Append(".ctor ");
+                    }
+
+                    appendTypeParameterTypeConstraints(typeParameter, builder);
+                }
+                else
+                {
+                    builder.Append('!');
+                }
+
+                builder.Append('T');
+                builder.Append(StringExtensions.GetNumeral(typeParameter.Ordinal));
```

Comment:
>builder.Append(StringExtensions.GetNumeral(typeParameter.Ordinal)); [](http://example.com/codeflow?start=16&length=67)

Is it really useful to add an ordinal at the end of a declaration? It feels like it is better to have two distinct helpers "add declaration" and "add reference" because they really have nothing in common in terms of behavior.

Reply:
I can split into two helpers. But both declaration and reference should have "T" and an ordinal, to conform to valid IL